src/sssom/util.py

```diff
@@ -299,6 +300,56 @@ def remove_mappings(self, msdf: "MappingSetDataFrame") -> None:
         self.df = self.df[self.df.columns.drop(list(self.df.filter(regex=r"_2")))]
         self.clean_prefix_map()
 
+    def propagate(self) -> List[str]:
+        """Propagate slot values from the set level down to individual records.
+
+        :return: The list of slots that were effectively propagated.
+        """
+        schema = SSSOMSchemaView()
+        propagated = []
+
+        for slot in schema.propagatable_slots:
+            if slot not in self.metadata:  # Nothing to propagate
+                continue
+            if slot in self.df.columns:  # Cannot propagate
+                continue
+
+            if schema.view.get_slot(slot).multivalued:
+                value = "|".join(self.metadata.pop(slot))
+            else:
+                value = self.metadata.pop(slot)
+
+            self.df[slot] = value
+            propagated.append(slot)
+
+        return propagated
+
+    def condense(self) -> List[str]:
+        """Condense record-level slot values to the set whenever possible.
+
+        :return: The list of slots that were effectively condensed.
+        """
+        schema = SSSOMSchemaView()
+        condensed = []
+
+        for slot in schema.propagatable_slots:
+            if slot not in self.df.columns:  # Nothing to condense
+                continue
+            values = self.df[slot].unique()
+            if len(values) > 1:  # Cannot condense
+                continue
+
+            if schema.view.get_slot(slot).multivalued:
+                value = values[0].split("|")
+            else:
+                value = values[0]
+
+            self.metadata[slot] = value
```

Comment:
Note to self before I forget: this is wrong, [as per the spec](https://mapping-commons.github.io/sssom/spec-formats-tsv/#condensation) condensation is only allowed if there is not already a value at the set level (or if the set-level value happens to be the same as the values for all mapping records), so here we are missing a check (does the `self.metadata` dict already have a value for the slot?).